### ISSUE TITLE
std::size_t to int conversion warnings in thrust::transform_reduce

### ISSUE THREAD

**Issue by endorph-soft** [NVIDIA/thrust#974]
I'm getting a conversion warning in thrust::transform_reduce.

In short, line 51 of transform_reduce.h defines size_type as:

`typedef typename iterator_traits<InputIt>::difference_type size_type;`

As expected, this is std::ptrdiff_t for a pointer.

This then calls reduce_n, which internally calls thrust::cuda_cub::detail::reduce_n_impl. That function takes a 32 bit int, not the 64 bit ptrdiff_t. I get a 50-line compiler warning as a result.

It would be nice if this did not happen. I don't believe I can fix it externally, without simply silencing the warning, which is not preferable.

**Comment by ToBeckmann** [comment]
I'm getting a similar warning, even in the simpler `thrust_reduce`.

I would really like to treat warnings as errors, and this is currently the only one left in our codebase...

Short test-case:
```C
#include <vector>
#include <thrust/device_vector.h>
#include <thrust/transform.h>

float host_ptr[6] = { 1., 0., 2., 2., 1., 3. };
float host_result = thrust::reduce(host_ptr, host_ptr + 6);  // compare https://thrust.github.io/doc/group__reductions.html

thrust::device_ptr<float> dev_ptr(host_ptr);
// everything fine up to here...

float dev_result_ptr = thrust::reduce(dev_ptr, dev_ptr + 6);  // but this compiles with warning: size_t <> int
```

I'm using CUDA 10.1 on MSVC 2015, which results in this command line (slightly shortened: paths):
`"[path]CUDA\v10.1\bin\nvcc.exe" -gencode=arch=compute_50,code=\"sm_50,compute_50\" --use-local-env -ccbin "[path]14.0\VC\bin\x86_amd64" -x cu  -I[path]  -I"[path]CUDA\v10.1\include" -I"[path]CUDA\v10.1\include" -maxrregcount=0  --machine 64 --compile -cudart static -arch=compute_50 -code=sm_50 -g   -D_CRT_SECURE_NO_WARNINGS -DNOMINMAX -DUNICODE -D_WINDLL -D_UNICODE -DUNICODE -Xcompiler "/EHsc /W3 /nologo /Od /FS /Zi  /MDd " "[path]]\reduce_demo.cu"`

I'm pasting only head and tail of the convoluted templatish warning:
```
1> reduce.h(939): warning C4244: 'argument': conversion from 'size_type' to 'int', possible loss of data
1> reduce.h(1017): note: see reference to function template instantiation 'T thrust::cuda_cub::detail::reduce_n_impl<thrust::cuda_cub::tag,InputIt,Size,T,BinaryOp>(thrust::cuda_cub::execution_policy<thrust::cuda_cub::tag> &,InputIt,Size,T,BinaryOp)' being compiled
1>          with
1>          [
1>              T=value_type,
1>              InputIt=thrust::device_ptr<float>,
1>              Size=size_type,
1>              BinaryOp=thrust::plus<float>
1>          ]
1>  reduce.h(1037): <...>
1>  reduce.h(1049):  <...>
1>  reduce.h(1061): <...>
1>  reduce.inl(41): <...>
1>  reduce.inl(152): <...>
1>  reduce_demo.cu(11): note: see reference to function template instantiation 'float thrust::reduce<thrust::device_ptr<float>>(InputIterator,InputIterator)' being compiled
1>          with
1>          [
1>              InputIterator=thrust::device_ptr<float>
1>          ]
1>C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include\thrust/system/cuda/detail/reduce.h(968): warning C4244: 'argument': conversion from 'size_type' to 'int', possible loss of data
1>```

**Comment by zhihaoy** [comment]
cub `Reduce` accept `int num_items`, but `Size` is passed in without casting.  `Size` may be produced by Thrust itself rather than controlled by users.  Similar warnings:
```
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include\thrust/system/cuda/detail/reduce.h(940): warning C4244: 'argument': conversion from 'Size' to 'int', possible loss of data [....vcxproj]
C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.1\include\thrust/system/cuda/detail/reduce.h(969): warning C4244: 'argument': conversion from 'Size' to 'int', possible loss of data [....vcxproj]
```

MSVC 19.23.28106.4

**Comment by brycelelbach** [comment]
This looks like a known issue with us using `int` in the CUB backend; we're working on fixing it.

**Comment by brycelelbach** [comment]
@griwes to confirm.

**Comment by griwes** [comment]
Yes, this is a part of that issue.

**Comment by allisonvacanti** [comment]
This should have been fixed by 500c4e07206fcde1d4060fe4397eed8b03953def, and some related warning are being fixed as part of #1359.

Please let me know if you're still seeing this warning with both of those patches present.
